cppfd/parallel_mesh.cpp

```diff
@@ -0,0 +1,89 @@
+#include "parallel_mesh.h"
+
+#include <iomanip>
+#include <iostream>
+#include <array>
+#include <map>
+#include <cmath> 
+
+ParallelMesh::ParallelMesh(uint64_t n_x, uint64_t n_y, double cell_size,
+			   uint16_t n_p, uint16_t n_q,
+			   double o_x, double o_y){
+
+  // set instance variables
+  this->n_cells_x = n_x;
+  this->n_cells_y = n_y;
+  this->n_blocks_x = n_p;
+  this->n_blocks_y = n_q;
+  this->h = cell_size;
+  this->origin = {o_x, o_y};
+  this->q_x = n_x / n_p;
+  this->q_y = n_y / n_q;
+  this->r_x = n_x % n_p;
+  this->r_y = n_y % n_q;
+  
+  // iterate over row (Y) major over mesh chunks
+  for (auto q = 0; q < n_q; q++){
+    // determine row height
+    auto n = (q < this->r_y) ? this->q_y + 1 : this->q_y;
+    
+    // initialze column (X) horizontal origin
+    o_x = this->get_origin()[0];
+
+    // iterate over column (X) minor over mesh chunks
+    for (auto p = 0; p < n_p; p++){
+      // determine column width
+      auto m = (p < this->r_x) ? this->q_x + 1 : this->q_x;
+
+      // create default mesh chunk boundary point types
+      std::map<uint8_t, PointTypeEnum> pt;
+      pt[0] = pt[1] = pt[3] = pt[4] = pt[7] = PointTypeEnum::GHOST;
+      pt[2] = pt[5] = pt[6] = PointTypeEnum::SHARED_OWNED;
+
+      // override outer boundary point types when applicable
+      if (p == 0)
+	pt[0] = pt[3] = pt[7] = PointTypeEnum::BOUNDARY;
+      if (p == n_p - 1)
+	pt[1] = pt[2] = pt[5] = PointTypeEnum::BOUNDARY;
+      if (q == 0)
+	pt[0] = pt[1] = pt[4] = PointTypeEnum::BOUNDARY;
+      if (q == n_q - 1)
+	pt[2] = pt[3] = pt[6] = PointTypeEnum::BOUNDARY;
+
+      // instantiate and store new mesh block
+      this->mesh_chunks.push_back(MeshChunk(m, n, this->h, pt, o_x, o_y));
```

Comment:
You can use `emplace_back` here if you want. It is slightly more optimized.

Reply:
Thanks for comments here, sorry I had missed yours @jstrzebonski (I find the GitHub review comments unwieldy to navigate at times). So yes indeed I ended up using `emplace` but that's finally on a `map` rather than a `vector` in order to allow for convenient retrieval of parallel mesh chunks by Cartesian coarse grid indexing.